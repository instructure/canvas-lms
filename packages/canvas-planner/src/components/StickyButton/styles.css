.root {
  box-sizing: border-box;
  display: block;
  border: none;
  color: var(--color);
  background-color: var(--background);
  padding: 0;
  font-size: var(--fontSize);
  font-weight: var(--fontWeight);
  font-family: var(--fontFamily);
  text-transform: var(--textTransform);
  line-height: var(--lineHeight);
  white-space: nowrap;
  cursor: pointer;
  user-select: none;
  touch-action: manipulation;
  appearance: none;
  transition: background-color 0.2s;
  outline: none;
  overflow: visible;
  border-bottom-left-radius: var(--borderRadius);
  border-bottom-right-radius: var(--borderRadius);
  position: fixed;

  &::before {
    content: "";
    box-sizing: border-box;
    width: calc(100% + 0.5rem);
    height: calc(100% + 0.5rem);
    border: var(--focusRingWidth) solid var(--focusRingColor);
    position: absolute;
    top: -0.25rem;
    left: -0.25rem;
    border-bottom-left-radius: var(--borderRadius);
    border-bottom-right-radius: var(--borderRadius);
    transform: scale(0.25);
    opacity: 0;
    transition: all 0.2s ease-out;
  }

  &:focus {
    &::before {
      opacity: 1;
      transform: scale(1);
    }
  }

  &:focus,
  &:hover {
    background-color: var(--backgroundHover);

    .icon {
      transform: translate3d(0, -0.0625rem, 0) scale(1.2);
    }
  }

  &[aria-disabled] {
    cursor: not-allowed;
    pointer-events: none;
    opacity: 0.5;
  }
}

.icon {
  display: block;
  font-size: 0.75rem;
  margin-left: var(--iconMargin);
  transform: translate3d(0, -0.0625rem, 0);
  transition: all 0.2s;
}

.direction--up,
.direction--down {
  .layout {
    padding-right: var(--hasIconRightPadding);
  }
}

.layout {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: var(--padding);
}
