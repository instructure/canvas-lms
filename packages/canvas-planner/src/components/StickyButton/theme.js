/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { darken } from '@instructure/ui-themeable/lib/utils/color';

export default function generator ({ borders, colors, typography, spacing }) {
  return {
    fontSize: typography.fontSizeXSmall,
    fontFamily: typography.fontFamily,
    fontWeight: typography.fontWeightNormal,
    color: colors.white,
    background: colors.brand,
    backgroundHover: darken(colors.brand, 5),
    padding: `0 ${spacing.small}`,
    textTransform: 'uppercase',
    lineHeight: spacing.medium,
    iconMargin: spacing.xxSmall,
    hasIconRightPadding: spacing.xSmall,
    borderRadius: borders.radiusMedium,
    focusRingWidth: borders.widthSmall,
    focusRingColor: colors.brand
  };
}

generator.canvas = function (variables) {
  return {
    background: variables["ic-brand-primary"],
    backgroundHover: darken(variables["ic-brand-primary"], 5),
    focusRingColor: variables["ic-brand-primary"],
  };
};
