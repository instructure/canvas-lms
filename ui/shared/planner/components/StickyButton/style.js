/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {darken} from '@instructure/ui-color-utils'
import {getThemeVars} from '../../getThemeVars'

export default function buildStyle() {
  /*
   * If the theme variables to be used when generating the styles below
   * are dependent on the actual theme in use, you can also pull out the
   * `key` property from the return from `getThemeVars()` and do a bit of
   * if or switch statement logic to get the result you want.
   */
  const {variables, key} = getThemeVars()

  let themeAdditionalStyles = {}
  switch (key) {
    case 'canvas':
      themeAdditionalStyles = {
        background: variables['ic-brand-primary'],
        backgroundHover: darken(variables['ic-brand-primary'], 5),
        focusRingColor: variables['ic-brand-primary'],
      }
      break
  }

  const classNames = {
    root: 'StickyButton-styles__root',
    icon: 'StickyButton-styles__icon',
    layout: 'StickyButton-styles__layout',
    newActivityButton: 'StickyButton-styles__newActivityButton',
    directionUp: 'StickyButton-styles__direction--up',
    directionDown: 'StickyButton-styles__direction--down',
  }

  const theme = {
    fontSize: variables.typography.fontSizeXSmall,
    fontFamily: variables.typography.fontFamily,
    fontWeight: variables.typography.fontWeightNormal,
    color: variables.colors.contrasts.white1010,
    background: variables.colors.contrasts.blue4570,
    backgroundHover: darken(variables.colors.contrasts.blue4570, 5),
    padding: `0 ${variables.spacing.small}`,
    textTransform: 'uppercase',
    lineHeight: variables.spacing.medium,
    iconMargin: variables.spacing.xxSmall,
    hasIconRightPadding: variables.spacing.xSmall,
    borderRadius: variables.borders.radiusMedium,
    focusRingWidth: variables.borders.widthSmall,
    focusRingColor: variables.colors.contrasts.blue4570,
    ...themeAdditionalStyles,
  }

  const css = `
  .${classNames.root} {
    box-sizing: border-box;
    display: block;
    border: none;
    color: ${theme.color};
    background-color: ${theme.background};
    padding: 0;
    font-size: ${theme.fontSize};
    font-weight: ${theme.fontWeight};
    font-family: ${theme.fontFamily};
    text-transform: ${theme.textTransform};
    line-height: ${theme.lineHeight};
    white-space: nowrap;
    cursor: pointer;
    user-select: none;
    touch-action: manipulation;
    appearance: none;
    transition: background-color 0.2s;
    outline: none;
    overflow: visible;
    border-bottom-left-radius: ${theme.borderRadius};
    border-bottom-right-radius: ${theme.borderRadius};
    position: fixed;
  }
  .${classNames.root}::before {
    content: "";
    box-sizing: border-box;
    width: calc(100% + 0.5rem);
    height: calc(100% + 0.5rem);
    border: ${theme.focusRingWidth} solid ${theme.focusRingColor};
    position: absolute;
    top: -0.25rem;
    /* the placement and radii are symetrical, so no need to replace left/right with start/end */
    left: -0.25rem;
    border-bottom-left-radius: ${theme.borderRadius};
    border-bottom-right-radius: ${theme.borderRadius};
    transform: scale(0.25);
    opacity: 0;
    transition: all 0.2s ease-out;
  }
  .${classNames.root}:focus::before {
    opacity: 1;
    transform: scale(1);
  }
  .${classNames.root}:focus, .${classNames.root}:hover {
    background-color: ${theme.backgroundHover};
  }
  .${classNames.root}:focus .${classNames.icon}, .${classNames.root}:hover .${classNames.icon} {
    transform: translate3d(0, -0.0625rem, 0) scale(1.2);
  }
  .${classNames.root}[aria-disabled] {
    cursor: not-allowed;
    pointer-events: none;
    opacity: 0.5;
  }

  .${classNames.icon} {
    display: block;
    font-size: 0.75rem;
    margin-inline-start: ${theme.iconMargin};
    transform: translate3d(0, -0.0625rem, 0);
    transition: all 0.2s;
  }

  .${classNames.directionUp} .${classNames.layout},
  .${classNames.directionDown} .${classNames.layout} {
    padding-inline-end: ${theme.hasIconRightPadding};
  }

  .${classNames.layout} {
    box-sizing: border-box;
    display: flex;
    align-items: center;
    width: 100%;
    height: 100%;
    padding: ${theme.padding};
  }

  .${classNames.newActivityButton} {
    inset-inline-end: 0;
    top: 100%;
    position: absolute;
  }
  `

  return {css, classNames, theme}
}
