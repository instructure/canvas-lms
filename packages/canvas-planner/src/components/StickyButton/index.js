/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react';
import classnames from 'classnames';
import themeable from '@instructure/ui-themeable/lib';
import { bool, func, node, number, string, oneOf } from 'prop-types';
import IconArrowUpSolid from 'instructure-icons/lib/Solid/IconArrowUpSolid';
import IconArrowDownLine from 'instructure-icons/lib/Line/IconArrowDownLine';

import styles from './styles.css';
import theme from './theme.js';

class StickyButton extends Component {
  static propTypes = {
    children: node.isRequired,
    onClick: func,
    disabled: bool,
    direction: oneOf(['none', 'up', 'down']),
    zIndex: number,
    offset: string,
    buttonRef: func,
  };

  static defaultProps = {
    direction: 'none',
    offset: '0',
  };

  handleClick = (e) => {
    const {
      disabled,
      onClick
    } = this.props;

    if (disabled) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };

  renderIcon () {
    const direction = this.props.direction;

    if (direction === 'up') {
      return <IconArrowUpSolid className={styles.icon} />;
    } else if (direction === 'down') {
      return <IconArrowDownLine className={styles.icon} />;
    } else {
      return null;
    }
  }

  render () {
    const {
      children,
      disabled,
      direction,
      zIndex,
      offset,
    } = this.props;

    const classes = {
      [styles.root]: true,
      [styles['direction--' + direction]]: direction !== 'none'
    };

    const style = {
      zIndex: (zIndex) ? zIndex : null,
      top: offset,
    };

    return (
      <button
        type="button"
        onClick={this.handleClick}
        className={classnames(classes)}
        style={style}
        aria-disabled={(disabled) ? 'true' : null}
        ref={this.props.buttonRef}
      >
        <span className={styles.layout}>
          {children}
          {this.renderIcon()}
        </span>
      </button>
    );
  }
}

export default themeable(theme, styles)(StickyButton);
