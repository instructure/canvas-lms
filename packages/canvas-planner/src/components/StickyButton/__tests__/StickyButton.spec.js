/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react';
import StickyButton from '../index';
import {shallow} from 'enzyme';
import IconArrowUpSolid from 'instructure-icons/lib/Solid/IconArrowUpSolid';
import IconArrowDownLine from 'instructure-icons/lib/Line/IconArrowDownLine';


it('renders', () => {
  const wrapper = shallow(<StickyButton>I am a Sticky Button</StickyButton>);
  expect(wrapper).toMatchSnapshot();
});

it('calls the onClick prop when clicked', () => {
  const fakeOnClick = jest.fn();
  const wrapper = shallow(
    <StickyButton onClick={fakeOnClick}>
      Click me
    </StickyButton>
  );

  wrapper.find('button').simulate('click');
  expect(fakeOnClick).toHaveBeenCalled();
});

it('does not call the onClick prop when disabled', () => {
  const fakeOnClick = jest.fn();
  const wrapper = shallow(
    <StickyButton onClick={fakeOnClick} disabled>
      Disabled button
    </StickyButton>
  );

  wrapper.find('button').simulate('click', {
    preventDefault() {},
    stopPropagation() {}
  });
  expect(fakeOnClick).not.toHaveBeenCalled();
});

it('renders the correct up icon', () => {
  const wrapper = shallow(
    <StickyButton direction="up">
      Click me
    </StickyButton>
  );
  expect(wrapper.find(IconArrowUpSolid)).toHaveLength(1);
});

it('renders the correct down icon', () => {
  const wrapper = shallow(
    <StickyButton direction="down">
      Click me
    </StickyButton>
  );
  expect(wrapper.find(IconArrowDownLine)).toHaveLength(1);
});
