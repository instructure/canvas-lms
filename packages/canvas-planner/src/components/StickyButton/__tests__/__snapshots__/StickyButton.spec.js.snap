// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<button
  aria-disabled={null}
  className=""
  onClick={[Function]}
  style={
    Object {
      "top": "0",
      "zIndex": null,
    }
  }
  type="button"
>
  <span>
    I am a Sticky Button
  </span>
</button>
`;
